chingu-voyages/v51-tier3-team-33
# Feature/oauth2.0

# Description
Implemented OAuth 2.0 with google credential provider, using nextauth.js; that is wired to the middleware of next.js . Now for every request made to the app, the url paths will be checked for protected paths. If a path is protected, it will be checked it there is a user session for that browser. If not, the user will be redirected to the login page; if they choose to continue google with google, they will be redirected to google OAuth consent screen. If consent is given and this app receives a callback from google with user google profile data, the app will automatically redirect the user to their originally intended destination.

## Related to the following Jira/Trello issue 
As a user, I want to provide my basic information (Email/ Username and password), so that Google can identify and authenticate me.

- Insert link to a relevant task/issue in Jira/Trello

## Main changes explained:
 Added the following files to the project: 
- middleware.ts: Exports NextAuth as the default module, allowing nextauth to intercept any request data
- auth.ts: Exports member functions relevant to nextauth. functions like handlers(GET, POST), serverside session management functions(signIn, signOut)
- auth.config.ts: Defines the route where to find the login and error pages, defines the authorize and signIn callback functions

Added an api route: `/api/auth/[...nextauth]/route.ts`
- This route exports the handlers(GET, POST) defined in auth.ts. the GET is used by client side pages, to request session data for a user already signed in. POST is used by the OAuth callbacks.

## How to test:
1. check into current branch
2. do `npm install` and `npm run dev` to run this PR locally or use `pnpm` instead of `npm` as it works best with `nextauth@beta`
3. visit the link http://localhost:3000, (3000 is the running port of you app instance)
4. click the button `go to protected route`
5. if no session available, you should be redirected to the login page. 
6. Click login with google
7. You should be redirected to google OAuth consent page. Consent to allow app to access your profile data.
8. the app will redirect you to your originally intended route `http://localhost:3000/protected`
9. to test sign out functionality, click signOut button, and you should be redirected back to the login page.
10. modifications can be done to redirect user to the home (landing page)

## Screenshots or videos of changes:

https://github.com/user-attachments/assets/ffcf8d74-ab7e-4d6e-8958-2b0a96474827

## Note:
Hope you enjoy reviewing the changes.
